### PR TITLE
Support `use`ing externally defined macros behind `#![feature(use_extern_macros)]`

### PR DESCRIPTION
**Pull request by jseyfried**
With `#![feature(use_extern_macros)]`,
 - A name collision between macros from different upstream crates is much less of an issue since we can `use` the macros in different submodules or rename with `as`.
 - We can reexport macros with `pub use`, so `#![feature(macro_reexport)]` is no longer needed.
 - These reexports are allowed in any module, so crates can expose a macro-modular interface.

If a macro invocation can resolve to both a `use` import and a `macro_rules!` or `#[macro_use]`, it is currently an ambiguity error.

r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc #35896, #35900

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -518,10 +518,12 @@ impl<'b> Resolver<'b> {
         module.populated.set(true)
     }
 
-    fn legacy_import_macro(&mut self, name: Name, def: Def, span: Span, allow_shadowing: bool) {
-        self.used_crates.insert(def.def_id().krate);
+    fn legacy_import_macro(
+        &mut self, name: Name, binding: &'b NameBinding<'b>, span: Span, allow_shadowing: bool,
+    ) {
```
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: formatting nit: prefer one arg per line

File: src/librustc_resolve/build_reduced_graph.rs [hunk 2]
```diff
@@ -338,27 +353,28 @@ impl<'b> Resolver<'b> {
 
     // Constructs the reduced graph for one variant. Variants exist in the
     // type and value namespaces.
-    fn build_reduced_graph_for_variant(&mut self,
-                                       variant: &Variant,
-                                       parent: Module<'b>,
-                                       vis: ty::Visibility) {
+    fn build_reduced_graph_for_variant(
+        &mut self, variant: &Variant, parent: Module<'b>, vis: ty::Visibility, expansion: Mark,
```
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: style nit - sig - here and below

**Review by nrc — APPROVED** [review]
r = me with the style nits addressed

**Pull request merged**